### ISSUE TITLE
switch polling to event-driven refresh

### ISSUE THREAD

**Issue by jelkosz** [oVirt/ovirt-web-ui#283]
This issue is more a material for discussion. It may bring large performance enhancement but not sure how much do we need it and how much is it worth doing it.

Our currently by far largest performance bottleneck is the periodic polling of the heavy /vms endpoint. The ideal solution for this would be to implement some kind of server push, but it is an extremely risky and big change to the engine core which will not happen (at least not in the near future).

We are not the only ones which are facing this same issue though - a similar problems are solved also in ManageIQ and it's oVirt provider. They came up with an elegant solution how to find a compromise which boosts performance significantly and is already present in the API. It is called targeted refresh (or the new graph refresh but this underlying concept is the same) and works like this:

- at first, they do a normal poll to get the data
- than they are only polling the 
  GET /ovirt-engine/api/event;from=<event id>
- this returns all events which are newer than the <event id> (meaning you can just periodically look for new events which have been added since last poll and the same events are never polled again). Other advantage is that even if the client disconnects, no events are lost.
- this events are the normal audit log events which are already there. Since the ManageIQ is based on this and they are complete enough for them they are certainly complete enough for us. But even if we did miss something, there is no problem to add a new audit log message at any point in the engine.
- after getting an event like: "VM tj-bootable-3 configuration was updated by admin@internal-authz." This event is than parsed and evaluated what has to be done. In this case for example by the VM name (tj-bootable-3) you find the ID (from the frontend DB) and do a poll of this one VM and nothing else.
- this way it is possible to eliminate almost completely the heavy polling and poll only one lightweight endpoint. And still, the result is a much more fluent user experience which is almost like having server push implemented.

It is obviously not a bullet-proof solution and also not beautiful. On the other hand, it is heavily used by ManageIQ and this fact brings lots of pressure to kinda maintain the format of the messages backward compatible. We could piggyback on this effort and also rely on it.

Please note that we can not use this so far because the ;from=<event id> does not work for user level API: [1]
If we decide to use it, we need to push for solving the [1] or fix it ourselfs. It is not a big problem, just needs to be done.

@mareklibra  what do you think? 

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1469055

**Comment by mareklibra** [comment]
@jelkosz , thanks for posting that.

Recently the issue might be for _admin_ users where the list of VMs might be huge (works fine for lower hundreds but might be in thousands). Anyway, recent bottleneck might be easily removed - see bellow.

How it works now:
- full data load at application start-up (with LocalStorrage caching of some objects, like icons)
  - the most expensive part is retrieving of disk-data
  - recently implemented scrolling/paging helps a lot from user perspective - the "full load" is effectively just a partial one
- targeted deep refresh of a single VM when its detail is opened
- periodic shallow refresh (/vms endpoint)
  - expensive or barely changed data are skipped
- once per longer time period, full data refresh is performed

What we might simply do with recent approach:
  - postpone disk-data retrieval (or similar) until VM detail is requested by user
    - the disk-part will "blink" when particular VM detail is open for the first time, no issue when re-opening
    - recently there's no use of this data on the landing VmsList page
  - remove the periodic but _rare_ full-data refresh
    - no issue for list of VMs
    - up-to-date single-VM data are still being retrieved when VM detail is accessed

With this just a __little change, performance will be much improved and sufficient__, I believe.
And we do not need to care about outdated data.

Anyway, the proposed use of `/event` endpoint is conceptually better - brings us closer to event-driven data refresh.
The more work with implementation pays off decreased server-side load.
The `/event` based data update can be combined with recent infinite-scroller functionality.

The key question is about comprehensiveness of the audit log messages. I can't say at this moment.
We need to avoid "combined" approach with occasional polling.

So I don't think it's a priority now, but definitely worth of trying. It _might_ pay off in the future.

**Comment by mykaul** [comment]
Without knowing much about the way you access the API, the various SDKs (specifically, Ruby and Python) have implemented compression, pipelining and multiplexing, all of which greatly improved the performance of fetching objects. Is any of these techniques implemented in the ovirt-web-ui?

What about link following feature?

**Comment by jelkosz** [comment]
- compression is a browser feature and is used
- pipelining and multiplexing is not used here AFAIK
- the discussion about link following have been started by me and Im eager to make use of it once the code lands in master :)

**Comment by jelkosz** [comment]
> The key question is about comprehensiveness of the audit log messages. I can't say at this moment.

If it is enough for ManageIQ than it is certainly enough for ovirt-web-ui
Also, adding new audil logs here and there is easy...

> We need to avoid "combined" approach with occasional polling.

+1

> So I don't think it's a priority now, but definitely worth of trying. It might pay off in the future.

sure, I just wanted to start a discussion

**Comment by michalskrivanek** [comment]
I do not think we have any feedback on performance yet. Seems to perform nice as is and this change is a bit risky. Hence flagging with "minor" for now

**Comment by gregsheremeta** [comment]
duplicate of #390
